internal/resolvers/twitter/tweet_loader.go

```diff
@@ -119,15 +140,192 @@ func (l *TweetLoader) Load(ctx context.Context, tweetID string, r *http.Request)
 		return resolver.Errorf("Twitter tweet API error: %s", err)
 	}
 
-	tweetData := buildTweetTooltip(tweetResp)
+	tooltipData := l.buildTweetTooltip(ctx, tweetResp, r)
+
 	var tooltip bytes.Buffer
-	if err := tweetTooltipTemplate.Execute(&tooltip, tweetData); err != nil {
+	if err := tweetTooltipTemplate.Execute(&tooltip, tooltipData); err != nil {
 		return resolver.Errorf("Twitter tweet template error: %s", err)
 	}
 
 	return &resolver.Response{
 		Status:    http.StatusOK,
 		Tooltip:   url.PathEscape(tooltip.String()),
-		Thumbnail: tweetData.Thumbnail,
+		Thumbnail: tooltipData.Thumbnail,
 	}, cache.NoSpecialDur, nil
 }
+
+func (l *TweetLoader) buildTweetTooltip(
+	ctx context.Context,
+	tweet *TweetApiResponse,
+	r *http.Request,
+) *tweetTooltipData {
+	data := &tweetTooltipData{}
+	data.Text = tweet.Text
+	data.Name = tweet.User.Name
+	data.Username = tweet.User.Username
+	data.Likes = humanize.Number(tweet.Likes)
+	data.Retweets = humanize.Number(tweet.Retweets)
+
+	// TODO: what time format is this exactly? can we move to humanize a la CreationDteRFC3339?
+	timestamp, err := time.Parse("Mon Jan 2 15:04:05 -0700 2006", tweet.Timestamp)
+	if err != nil {
+		data.Timestamp = ""
+	} else {
+		data.Timestamp = humanize.CreationDateTime(timestamp)
+	}
+
+	data.Thumbnail = l.buildThumbnailURL(ctx, tweet, r)
+
+	return data
+}
+
+func (l *TweetLoader) buildThumbnailURL(
+	ctx context.Context,
+	tweet *TweetApiResponse,
+	r *http.Request,
+) string {
+	log := logger.FromContext(ctx)
+
+	numMedia := len(tweet.Entities.Media)
+	if numMedia == 1 {
+		// If tweet contains exactly one image, it will be used as thumbnail
+		return tweet.Entities.Media[0].Url
+	}
+
+	// More than one media item, need to compose a thumbnail
+	thumb, err := composeThumbnail(ctx, tweet.Entities.Media)
+	if err != nil {
+		log.Errorw("Couldn't compose Twitter collage",
+			"err", err,
+		)
+		return ""
+	}
+
+	outputBuf, metaData, err := thumb.ExportNative()
+	if err != nil {
+		log.Errorw("Couldn't export Twitter collage thumbnail",
+			"err", err,
+		)
+		return ""
+	}
+
+	parentKey := l.tweetCacheKeyProvider.CacheKey(ctx, tweet.ID)
+	collageKey := buildCollageKey(tweet.ID)
+	contentType := utils.MimeType(metaData.Format)
+
+	err = l.collageCache.Insert(ctx, collageKey, parentKey, outputBuf, contentType)
+	if err != nil {
+		log.Errorw("Couldn't insert Twitter collage thumbnail into cache",
+			"err", err,
+		)
+		return ""
+	}
+
+	return utils.FormatGeneratedThumbnailURL(l.baseURL, r, collageKey)
+}
+
+func composeThumbnail(
```

Comment:
With this being a free standing function, it needs to be able to access `cfg` from a gloal scope. It should instead, if possible, be made into a TweetLoader function so it can access cfg from itself. This will make the code more testable in the future

Reply:
Addressed in 4ead9b9